### PR TITLE
Stop skewing inference in ?'s desugaring

### LINKED ISSUE #39216: Tracking issue for future-incompatibility lint `resolve_trait_on_defaulted_unit`
This is the **summary issue** for the `resolve_trait_on_defaulted_unit` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made. For more information on the policy around future-compatibility warnings, see our breaking change policy guidelines.

### What is this lint about

Ordinarily, when a user doesn't specify the type of an expression and the type cannot be inferred, rust will raise an error. For example, consider this code:

    let _ = Default::default();

Because we haven't specified the type of `_` the compiler doesn't know what type of value the user is asking for a default of. And so we get this error:

```rust
error[E0282]: unable to infer enough type information about `Self`
 --> example.rs:2:13
  |
2 |     let _ = Default::default();
  |             ^^^^^^^^^^^^^^^^ cannot infer type for `Self`
  |
  = note: type annotations or generic parameter binding required

error: aborting due to previous error
```

However, due to an unfortunate quirk in Rust's type inference algorithms it is sometimes possible to sneak situations like this past the compiler. In these cases, the unspecified type is defaulted to `()`. For an example of this, you can try deserializing an unspecified type with serde:

    let _ = Deserialize::deserialize(foo)?;

In this case, the code will compile and will deserialize a value of type `()`.

This behaviour is set to change with the eventual rolling-out of `feature(never_type)`. Where defaulting is still used types will instead default to `!`. In the best case, code that currently relies on unspecified types defaulting to `()` will stop compiling. In the worst case, code will continue to compile but may execute differently, as in the above example where a `!` will be deserialised instead.

The `resolve_trait_on_defaulted_unit` warning is raised wherever the compiler thinks your program's behaviour may depend on the current defaulting rules.

### How to fix this warning/error

Be specific about what type you're using. In the serde example above this could be done by simply adding a type annotation:

    let _: () = Deserialize::deserialize(foo)?;

### Current status
- [x] https://github.com/rust-lang/rust/pull/39009 introduces the `resolve_trait_on_defaulted_unit` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/42894 makes the `resolve_trait_on_defaulted_unit` lint deny-by-default
- [ ] PR ? makes the `resolve_trait_on_defaulted_unit` lint a hard error

### LINKED ISSUE #51125: Defaulted unit types no longer error out (regression?)
This currently compiles (on stable and nightly). Till 1.25, it would trigger a [lint](https://github.com/rust-lang/rust/issues/39216) because it has inference default to `()` instead of throwing an error.

```rust
fn main() {}

struct Err;
fn load<T: Default>() -> Result<T, Err> {
    Ok(T::default())
}
fn foo() -> Result<(), Err> {
    let val = load()?; // defaults to ()
    Ok(())
}
```

([playpen](https://play.rust-lang.org/?gist=6f7cf9dafb9d1651b659c1f029413fc5&version=nightly&mode=debug))

That lint indicates that it would become a hard error in the future, but it's not erroring. It seems like a bunch of this was changed when we [stabilized `!`](https://github.com/rust-lang/rust/issues/48950).

That issue says

> Type inference will now default unconstrained type variables to `!` instead of `()`. The [`resolve_trait_on_defaulted_unit`](https://github.com/rust-lang/rust/issues/39216) lint has been retired. An example of where this comes up is if you have something like:

Though this doesn't really _make sense_, this looks like a safe way to produce nevers, which should, in short, never happen. It seems like this is related to https://github.com/rust-lang/rust/issues/40801 -- but that was closed as it seems to be a more drastic change.

Also, if you print `val`, it's clear that the compiler thought it was a unit type. This seems like one of those cases where attempting to observe the situation changes it.

We _should_ have a hard error here, this looks like a footgun, and as @SimonSapin mentioned has broken some unsafe code already. We had an upgrade lint for the hard error and we'll need to reintroduce it for a cycle or two since we've had a release without it. AFAICT this is a less drastic change than #40801, and we seem to have _intended_ for this to be a hard error before so it probably is minor enough that it's fine.

cc @nikomatsakis @eddyb 

h/t @spacekookie for finding this

### PR DESCRIPTION
**Pull request by WaffleLapkin**
**NB**: this is a breaking change (although arguably a bug fix) and as such shouldn't be taken lightly.

This changes `expr?`'s desugaring like so (simplified, see code for more info):
```rust
// old
match expr {
    Ok(val) => val,
    Err(err) => return Err(err),
}

// new
match expr {
    Ok(val) => val,
    Err(err) => core::convert::absurd(return Err(err)),
}

// core::convert
pub const fn absurd<T>(x: !) -> T { x }
```

This prevents `!` from the `return` from skewing inference:
```rust
// previously: ok (never type spontaneous decay skews inference, `T = ()`)
// with this pr: can't infer the type for `T`
Err(())?;
```

Fixes #51125
Closes #39216

### CODE REVIEW

File: library/core/src/convert/mod.rs [hunk 1]
```diff
@@ -105,6 +105,66 @@ pub const fn identity<T>(x: T) -> T {
     x
 }
 
+/// Converts [`!`] (the never type) to any type.
+///
+/// This is possible because `!` is uninhabited (has no values), so this function can't actually
+/// be ever called at runtime.
+///
+/// Even though `!` can be coerced to any type implicitly anyway (and indeed this function
```
Comment (nnethercote) [inline library/core/src/convert/mod.rs:113]: "function is implemented"

File: library/core/src/convert/mod.rs [hunk 2]
```diff
@@ -105,6 +105,66 @@ pub const fn identity<T>(x: T) -> T {
...
+/// ```compile_fail,E0282
+/// use core::convert::absurd;
+///
+/// let x: Result<_, ()> = Err(());
+/// let y = match x { //~ error[E0282]: type annotations needed
+///     Ok(v) => v,
+///
+///     // the call to `absurd` *is* unreachable, but it's still important for type check reasons
+///     #[allow(unreachable_code)]
+///     Err(()) => absurd(return),
```
Comment (nnethercote) [inline library/core/src/convert/mod.rs:142]: Why is this unreachable? `x` equals `Err(())`...

**Comment by nnethercote** [comment]
I have skimmed this PR and gave minor comments, but this kind of subtle lang/libs/type system change is the kind of thing I am perfectly *not* suited to reviewing. The "needs-fcp" and "I-lang-nominated" labels have been added, so this clearly will get some scrutiny via other channels, but I will take a guess at a suitable reviewer and reassign anyway...

r? @cuviper

**Comment by cuviper** [comment]
Is `convert::absurd` meant to be permanently unstable? If not, then it will need libs-api review.

**Comment by WaffleLapkin** [comment]
@cuviper I think it could plausibly become stable, at least if/when `!` is stabilized

File: library/core/src/convert/mod.rs [hunk 3]
```diff
@@ -105,6 +105,66 @@ pub const fn identity<T>(x: T) -> T {
...
+///
+/// ```rust
+/// let x: Result<_, ()> = Err(());
+/// let y = match x {
+///     Ok(v) => v,
+///     Err(()) => return,
+/// };
+/// ```
+///
+/// This is a bit unexpected, because the type of `y` is seemingly unbound (indeed, it can be any
```
[the remark this answers is not in the document]
Comment (traviscross) [reply library/core/src/convert/mod.rs:127]: The word "unbound" is correct here.  The type is "not bound" or "without a binding".  The word "unbounded" means something more like "limitless", which one could make a narrow argument for, but it seems connotatively wrong to my ear.

File: library/core/src/convert/mod.rs (hunk 2 above)
Comment (traviscross) [reply library/core/src/convert/mod.rs:142]: The `return` is evaluated before the call to `absurd`, so we can never reach that call.

**Comment by cuviper** [comment]
OK, well the actual implementation of `absurd` is not interesting, so let's get an API person to look...

r? libs-api

**Comment by chorman0773** [comment]
FTR, I've commonly used `Err(e)?` for `return Err(e)` - if this will cause issues in `if`/`else` blocks particularily, that would be a potential issue for my code.

**Comment by WaffleLapkin** [comment]
> it made me wonder whether it would be feasible to change return in general to be a free type variable instead of `!`?

@scottmcm I'm not sure. I don't think it's unfeasible, but it sure is harder than this.

The issues are:
- Need to add fallback for those type variables too, so that `return;` works
- `{ return; }` (which is currently `!` even though there is `;`) needs to be special cased in a different way
- Will break strictly more things

I'm not sure if this is a good idea or not. It's kinda weird.

**Comment by WaffleLapkin** [comment]
> I presume the next step is a crater run to evaluate the impact?

@Amanieu yes.

**Comment by WaffleLapkin** [comment]
@chorman0773 can you explain what you mean by "if this will cause issues in `if`/`else` blocks particularily"?

This PR makes all `Err(..)?` be compile time errors, the only way to avoid this is to specify the type with a turbofish (`Err::<T, _>(..)?`).

**Comment by chorman0773** [comment]
I generally write
```rust
if some_error_case{
     Err(/*error value*/)?
}
```
(or equivalent when using `match`), particularily when the error value (or the condition check) is more complicated - I find it nicer to write in many cases.

**Comment by traviscross** [comment]
As **scottmcm** mentioned, we discussed this in triage today.  On the one hand, people were hesitant to block incremental progress, but on the other, people were hesitant to add a special case if we could address a more general case.  There was, I would say, appetite for taking a bigger bite here, but people were uncertain if there were any bigger bites that were feasible other than those discussed to support the never type generally, such as disabling fallback to `()`.

In terms of next steps, we wanted to see an answer about the pros and cons of doing this for `return` generally, which @WaffleLapkin has now [answered](https://github.com/rust-lang/rust/pull/122412#issuecomment-2010480706) (please let us know @WaffleLapkin if you have further thoughts here), and we wanted to see the results of the crater run that we know that @WaffleLapkin is working to make happen.

When taking this back up, in addition to those details, we wanted to specifically consider how this incremental step may be addressing known footguns with unsafe code such as that in #51125.

**Comment by WaffleLapkin** [comment]
@chorman0773 this exact case actually will still work:
```rust
if some_error_case{
     Err(/*error value*/)?
}
```
else-less `if` requires the block to be of type `()`, so the type can be inferred. But if you add a `;` or there is no way to infer the type, it will fail.

**Comment by WaffleLapkin** [comment]
@craterbot run start=master#385fa9d845dd326c6bbfd58c22244215e431948a end=try#645bb72776a6a56a1a8f52631a44bd082b2ba509 mode=check-only name=absurd-question-mark-desugar crates=https://crater-reports.s3.amazonaws.com/no-never-type-fallback/retry-regressed-list.txt

(using https://github.com/rust-lang/rust/pull/122955 regressions as the crate list, since this can only break things that https://github.com/rust-lang/rust/pull/122955 broke)

**Comment by Amanieu** [comment]
r? compiler

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in the triage meeting today.  Some people liked this change, and most liked the spirit of it, but there were some concerns.  In particular, the many cases like this:

```diff
fn result() -> Result<bool, ()> {
-    Err(())?;
+    Err::<(), _>(())?;
     Ok(true)
}
```

...resulted in much discussion.  We did not reach consensus here other than that probably no consensus is possible at this time without a revised proposal, so we can unnominate until and unless such a revised proposal is available.

**Pull request closed without merging**